src/Symfony/Component/Yaml/Parser.php

```diff
@@ -286,7 +286,11 @@ public function parse($value, $exceptionOnInvalidType = false, $objectSupport =
             }
 
             if ($isRef) {
-                $this->refs[$isRef] = end($data);
+                if (!isset($key)) {
```

Comment:
detecting whether ``$key`` is set to detect whether we are in a sequence or a mapping is not very readable IMO (I had to read the parser code 3 times before being sure that there was no flow where the ``$key`` variable could come from a previous iteration, as we never reset it).
I suggest the following change:

```php
$this->refs[$isRef] = 'mapping' === $context ? $data[$key] : end($data);
```

This would identify much better that the case where we read based on the key is for mappings